src/observableInterfaces.ts

```diff
@@ -316,48 +322,31 @@ export type RecordValue<T> = T extends Record<string, infer t> ? t : never;
 export type ArrayValue<T> = T extends Array<infer t> ? t : never;
 export type ObservableValue<T> = T extends Observable<infer t> ? t : never;
 
-// This converts the state object's shape to the field transformer's shape
-// TODO: FieldTransformer and this shape can likely be refactored to be simpler
-declare type ObjectKeys<T> = Pick<
-    T,
-    {
-        [K in keyof T]-?: K extends string
-            ? T[K] extends Record<string, any>
-                ? T[K] extends any[]
-                    ? never
-                    : K
-                : never
-            : never;
-    }[keyof T]
->;
-declare type DictKeys<T> = Pick<
-    T,
-    {
-        [K in keyof T]-?: K extends string ? (T[K] extends Record<string, Record<string, any>> ? K : never) : never;
-    }[keyof T]
->;
-declare type ArrayKeys<T> = Pick<
-    T,
-    {
-        [K in keyof T]-?: K extends string | number ? (T[K] extends any[] ? K : never) : never;
-    }[keyof T]
->;
-export declare type FieldTransforms<T> =
+type FilterKeysByValue<T, U> = {
+    [K in keyof T]: T[K] extends U ? K & (string | number) : never;
+}[keyof T];
+
+type ObjectKeys<T> = Exclude<FilterKeysByValue<T, Record<string, any>>, FilterKeysByValue<T, any[]>>;
+type DictKeys<T> = FilterKeysByValue<T, Record<string, Record<string, any>>>;
+type ArrayKeys<T> = FilterKeysByValue<T, any[]>;
+type PrimitiveKeys<T> = FilterKeysByValue<T, Primitive>;
+
+export type FieldTransforms<T> =
```

Comment:
I didn't look into simplifying FieldTransforms yet, as I didn't check how the types are being used.
I think we should (maybe in a separate PR) move the types that have to do with persistence into a separate file, thoughts?

Reply:
Agreed. That file got too long and it would be great to separate the persistence types.